src/request_queue.js

```diff
@@ -135,6 +158,12 @@ export class RequestQueue {
         this.inProgressCount = 0;
         this.queryQueueHeadPromise = null;
 
+        // Caching requests to avoid duplicite addRequest() calls.
+        // Key is local cacheId and value requestId.
+        this.requestsCache = {};
+        // Cache of handled requests. Key is cacheId and value is true if request was already handled.
+        this.handledRequestsCache = {};
```

Comment:
Is there any point of handledRequestsCache? We can just store everything to requestsCache, including info whether the request was already handled. Also we should probably use LruCache with e.g. length of 1M - otherwise for extra large workloads the memory will overflow.